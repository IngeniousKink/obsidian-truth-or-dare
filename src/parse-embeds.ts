import { PhrasingContent } from "mdast";
import { Image, Text } from "node_modules/mdast-util-from-markdown/lib/index.js";

function createTextNode(content: PhrasingContent, newValue: string): Text {
    return {
        type: 'text',
        value: newValue,
        position: content.position
    };
}

function createImageContent(content: PhrasingContent, newValue: string): Image {
    return {
        type: 'image',
        url: newValue,
        position: content.position
    };
}

function processImageLink(
    content: Text,
    regexMatch: RegExpExecArray,
    previousIndex: number,
    processedContent: PhrasingContent[]
): number {
    if (regexMatch.index !== previousIndex) {
        processedContent.push(
            createTextNode(
                content,
                content.value.substring(previousIndex, regexMatch.index)
            )
        );
    }

    processedContent.push(createImageContent(content, regexMatch[1]));

    // Update the previousIndex to the end of the matched string
    return regexMatch.index + regexMatch[0].length;
}


function processMarkdownContent(content: PhrasingContent): PhrasingContent[] {
    const processedContent: PhrasingContent[] = [];
    const imageLinkRegex = /!\[\[(.*?)]]/g;
    let regexMatch;
    let previousIndex = 0;

    if (content.type !== 'text') return [content];

    while ((regexMatch = imageLinkRegex.exec(content.value)) !== null) {
        previousIndex = processImageLink(
            content,
            regexMatch,
            previousIndex,
            processedContent
        );
    }

    if (previousIndex !== content.value.length) {
        processedContent.push(
            createTextNode(
                content,
                content.value.substring(previousIndex)
            )
        );
    }

    return processedContent;
}

export function parseEmbeds(input: PhrasingContent[]): PhrasingContent[] {
    return input.flatMap(processMarkdownContent);
}
