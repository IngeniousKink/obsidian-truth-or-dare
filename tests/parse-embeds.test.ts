import { parseEmbeds } from '../src/parse-embeds.js';

describe('parse-embeds', () => {
  test('embed at start of textNode', () => {
    const result = parseEmbeds([{
      "type": "text",
      "value": "![[a_video.mp4]] and here goes some text",
      "position": {
        "start": {
          "line": 4,
          "column": 3,
          "offset": 72
        },
        "end": {
          "line": 4,
          "column": 40,
          "offset": 109
        }
      }
    }]);
    expect(result).toMatchSnapshot();
    expect(result.length).toBe(2);
  });

  test('embed at end of textNode', () => {
    const result = parseEmbeds([{
      "type": "text",
      "value": "And here is another video: ![[another_video.mp4]]",
      "position": {
        "start": {
          "line": 4,
          "column": 3,
          "offset": 72
        },
        "end": {
          "line": 4,
          "column": 40,
          "offset": 109
        }
      }
    }]);
    expect(result).toMatchSnapshot();
    expect(result.length).toBe(2);
  });

  test('embed in the middle of a textNode', () => {
    const result = parseEmbeds([{
      "type": "text",
      "value": "First some text ![[then_a_video.mp4]] then some more text",
      "position": {
        "start": {
          "line": 4,
          "column": 3,
          "offset": 72
        },
        "end": {
          "line": 4,
          "column": 40,
          "offset": 109
        }
      }
    }]);
    expect(result).toMatchSnapshot();
    expect(result.length).toBe(3);
  });

  test('two embeds', () => {
    const result = parseEmbeds([{
      "type": "text",
      "value": "text ![[video1.mp4]] and ![[picture2.webp]] the end",
      "position": {
        "start": {
          "line": 4,
          "column": 3,
          "offset": 72
        },
        "end": {
          "line": 4,
          "column": 40,
          "offset": 109
        }
      }
    }]);
    expect(result).toMatchSnapshot();
    expect(result.length).toBe(5);
  });


  test('embed with size info', () => {
    const result = parseEmbeds([{
      "type": "text",
      "value": "Here's an embedded picture: ![[a_picture.webp|100]]",
      "position": {
        "start": {
          "line": 3,
          "column": 3,
          "offset": 18
        },
        "end": {
          "line": 3,
          "column": 54,
          "offset": 69
        }
      }
    }]);
    expect(result).toMatchSnapshot();
    expect(result.length).toBe(2);
  });


});



