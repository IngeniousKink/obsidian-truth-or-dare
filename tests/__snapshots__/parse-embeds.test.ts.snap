// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse-embeds embed at end of textNode 1`] = `
[
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": "And here is another video: ",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "image",
    "url": "another_video.mp4",
  },
]
`;

exports[`parse-embeds embed at start of textNode 1`] = `
[
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "image",
    "url": "a_video.mp4",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": " and here goes some text",
  },
]
`;

exports[`parse-embeds embed in the middle of a textNode 1`] = `
[
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": "First some text ",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "image",
    "url": "then_a_video.mp4",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": " then some more text",
  },
]
`;

exports[`parse-embeds embed with size info 1`] = `
[
  {
    "position": {
      "end": {
        "column": 54,
        "line": 3,
        "offset": 69,
      },
      "start": {
        "column": 3,
        "line": 3,
        "offset": 18,
      },
    },
    "type": "text",
    "value": "Here's an embedded picture: ",
  },
  {
    "position": {
      "end": {
        "column": 54,
        "line": 3,
        "offset": 69,
      },
      "start": {
        "column": 3,
        "line": 3,
        "offset": 18,
      },
    },
    "type": "image",
    "url": "a_picture.webp|100",
  },
]
`;

exports[`parse-embeds two embeds 1`] = `
[
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": "text ",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "image",
    "url": "video1.mp4",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": " and ",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "image",
    "url": "picture2.webp",
  },
  {
    "position": {
      "end": {
        "column": 40,
        "line": 4,
        "offset": 109,
      },
      "start": {
        "column": 3,
        "line": 4,
        "offset": 72,
      },
    },
    "type": "text",
    "value": " the end",
  },
]
`;
